File: internal/insights/error_hotspots/hotspots.go
package error_hotspots

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"
	"time"
)

type errorAccumulator struct {
	occurrences uint32
	sessions    map[string]struct{}
	lastSeenMs  float64
}

func scanSessionHotspots(
	path, sessionID string,
	accumulator map[[2]string]*errorAccumulator,
) {
	f, err := os.Open(path)
	if err != nil {
		return
	}
	defer f.Close()

	inFlight := make(map[string]toolCall)
	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 64*1024), 4*1024*1024)
	for scanner.Scan() {
		line := scanner.Text()
		if strings.TrimSpace(line) == "" {
			continue
		}
		var entry rawEntry
		if json.Unmarshal([]byte(line), &entry) != nil {
			continue
		}
		tsMs := 0.0
		if entry.Timestamp != nil {
			if ms, ok := parseTimestampMs(*entry.Timestamp); ok {
				tsMs = ms
			}
		}
		if entry.Message == nil || entry.Message.Content == nil {
			continue
		}
		role := ""
		if entry.Message.Role != nil {
			role = *entry.Message.Role
		}
		var blocks []json.RawMessage
		if json.Unmarshal(*entry.Message.Content, &blocks) != nil {
			continue
		}

		switch role {
		case "assistant":
			for _, b := range blocks {
				var block struct {
					Type string `json:"type"`
					ID   string `json:"id"`
					Name string `json:"name"`
				}
				if json.Unmarshal(b, &block) != nil || block.Type != "tool_use" {
					continue
				}
				if block.ID == "" {
					continue
				}
				name := block.Name
				if name == "" {
					name = "unknown"
				}
				inFlight[block.ID] = toolCall{toolName: name}
			}
		case "user":
			for _, b := range blocks {
				var block struct {
					Type      string           `json:"type"`
					ToolUseID string           `json:"tool_use_id"`
					IsError   *bool            `json:"is_error"`
					Content   *json.RawMessage `json:"content"`
				}
				if json.Unmarshal(b, &block) != nil || block.Type != "tool_result" {
					continue
				}
				call, ok := inFlight[block.ToolUseID]
				if !ok {
					continue
				}
				delete(inFlight, block.ToolUseID)

				if block.IsError == nil || !*block.IsError {
					continue
				}
				resultText := ""
				if block.Content != nil {
					resultText = toolResultText(*block.Content)
				}
				prefix := normalizeErrorPrefix(resultText)
				if prefix == "" {
					continue
				}
				key := [2]string{call.toolName, prefix}
				acc := accumulator[key]
				if acc == nil {
					acc = &errorAccumulator{sessions: make(map[string]struct{})}
					accumulator[key] = acc
				}
				acc.occurrences++
				acc.sessions[sessionID] = struct{}{}
				if tsMs > acc.lastSeenMs {
					acc.lastSeenMs = tsMs
				}
			}
		}
	}
}

// ComputeErrorHotspots scans sessions and returns repeated tool errors.
// Mirrors hotspots.rs::compute_error_hotspots.
func ComputeErrorHotspots(projectID string, days, minOccurrences uint32) (*ErrorHotspotsResponse, error) {
	projectDir, err := resolveProjectDir(projectID)
	if err != nil {
		return nil, err
	}
	if days < 1 {
		days = 1
	}
	if days > 90 {
		days = 90
	}
	if minOccurrences < 2 {
		minOccurrences = 2
	}

	nowMs := float64(time.Now().UnixMilli())
	cutoffMs := nowMs - float64(days)*86_400_000.0

	entries, err := os.ReadDir(projectDir)
	if err != nil {
		return nil, fmt.Errorf("read dir: %w", err)
	}

	accumulator := make(map[[2]string]*errorAccumulator)
	var scannedSessions uint32

	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".jsonl") {
			continue
		}
		info, err := entry.Info()
		if err != nil {
			continue
		}
		if float64(info.ModTime().UnixMilli()) < cutoffMs {
			continue
		}
		sessionID := strings.TrimSuffix(entry.Name(), ".jsonl")
		scannedSessions++
		scanSessionHotspots(projectDir+"/"+entry.Name(), sessionID, accumulator)
	}

	var hotspots []RepeatedToolError
	for key, acc := range accumulator {
		if acc.occurrences < minOccurrences {
			continue
		}
		sessionIDs := make([]string, 0, len(acc.sessions))
		for sid := range acc.sessions {
			sessionIDs = append(sessionIDs, sid)
		}
		sort.Strings(sessionIDs)
		hotspots = append(hotspots, RepeatedToolError{
			ToolName:     key[0],
			ErrorPrefix:  key[1],
			Occurrences:  acc.occurrences,
			SessionCount: uint32(len(sessionIDs)),
			SessionIDs:   sessionIDs,
			LastSeenMs:   acc.lastSeenMs,
		})
	}
	sort.Slice(hotspots, func(i, j int) bool {
		return hotspots[i].Occurrences > hotspots[j].Occurrences
	})
	if hotspots == nil {
		hotspots = []RepeatedToolError{}
	}

	return &ErrorHotspotsResponse{
		RepeatedErrors:  hotspots,
		ScannedSessions: scannedSessions,
	}, nil
}

File: src-tauri/src/insights/error_hotspots/hotspots.rs
use std::collections::HashMap;
use std::io::{BufRead, BufReader};
use std::path::Path;

use serde::{Deserialize, Serialize};

use super::shared::{
    normalize_error_prefix, parse_timestamp_ms, resolve_project_dir, tool_result_text, RawEntry,
    ToolCall,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RepeatedToolError {
    pub tool_name: String,
    pub error_prefix: String,
    pub occurrences: u32,
    pub session_count: u32,
    pub session_ids: Vec<String>,
    pub last_seen_ms: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ErrorHotspotsResponse {
    pub repeated_errors: Vec<RepeatedToolError>,
    pub scanned_sessions: u32,
}

#[derive(Default)]
struct ErrorAccumulator {
    occurrences: u32,
    sessions: std::collections::HashSet<String>,
    last_seen_ms: f64,
}

fn scan_session(
    path: &Path,
    session_id: &str,
    accumulator: &mut HashMap<(String, String), ErrorAccumulator>,
) -> Option<()> {
    let file = std::fs::File::open(path).ok()?;
    let reader = BufReader::with_capacity(64 * 1024, file);
    let mut in_flight: HashMap<String, ToolCall> = HashMap::new();

    for line in reader.lines().map_while(Result::ok) {
        if line.trim().is_empty() {
            continue;
        }
        let entry: RawEntry = match serde_json::from_str(&line) {
            Ok(e) => e,
            Err(_) => continue,
        };
        let ts_ms = entry
            .timestamp
            .as_deref()
            .and_then(parse_timestamp_ms)
            .unwrap_or(0.0);
        let msg = match entry.message {
            Some(m) => m,
            None => continue,
        };
        let content = match msg.content {
            Some(c) => c,
            None => continue,
        };
        let blocks = match content.as_array() {
            Some(a) => a,
            None => continue,
        };

        match msg.role.as_deref() {
            Some("assistant") => {
                for block in blocks {
                    if block.get("type").and_then(|v| v.as_str()) != Some("tool_use") {
                        continue;
                    }
                    let id = match block.get("id").and_then(|v| v.as_str()) {
                        Some(s) => s.to_string(),
                        None => continue,
                    };
                    let name = block
                        .get("name")
                        .and_then(|v| v.as_str())
                        .unwrap_or("unknown")
                        .to_string();
                    in_flight.insert(id, ToolCall { tool_name: name });
                }
            }
            Some("user") => {
                for block in blocks {
                    if block.get("type").and_then(|v| v.as_str()) != Some("tool_result") {
                        continue;
                    }
                    let id = match block.get("tool_use_id").and_then(|v| v.as_str()) {
                        Some(s) => s.to_string(),
                        None => continue,
                    };
                    let call = match in_flight.remove(&id) {
                        Some(c) => c,
                        None => continue,
                    };
                    let is_error = block
                        .get("is_error")
                        .and_then(|v| v.as_bool())
                        .unwrap_or(false);
                    if !is_error {
                        continue;
                    }
                    let result_text = block
                        .get("content")
                        .map(tool_result_text)
                        .unwrap_or_default();
                    let prefix = normalize_error_prefix(&result_text);
                    if prefix.is_empty() {
                        continue;
                    }
                    let key = (call.tool_name, prefix);
                    let acc = accumulator.entry(key).or_default();
                    acc.occurrences += 1;
                    acc.sessions.insert(session_id.to_string());
                    if ts_ms > acc.last_seen_ms {
                        acc.last_seen_ms = ts_ms;
                    }
                }
            }
            _ => {}
        }
    }
    Some(())
}

pub fn compute_error_hotspots(
    project_id: &str,
    days: u32,
    min_occurrences: u32,
) -> Result<ErrorHotspotsResponse, String> {
    let project_dir = resolve_project_dir(project_id)?;
    let days = days.clamp(1, 90);
    let min_occurrences = min_occurrences.max(2);

    let now_ms = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map_err(|e| e.to_string())?
        .as_millis() as f64;
    let cutoff_ms = now_ms - (days as f64) * 86_400_000.0;

    // Go's os.ReadDir returns entries sorted by filename; mirror that.
    let mut entries: Vec<_> = std::fs::read_dir(&project_dir)
        .map_err(|e| e.to_string())?
        .flatten()
        .collect();
    entries.sort_by_key(|e| e.file_name());

    let mut accumulator: HashMap<(String, String), ErrorAccumulator> = HashMap::new();
    let mut scanned_sessions: u32 = 0;

    for entry in entries {
        let fname = entry.file_name();
        let fname = fname.to_string_lossy();
        if !fname.ends_with(".jsonl") {
            continue;
        }
        let modified_ms = entry
            .metadata()
            .ok()
            .and_then(|m| m.modified().ok())
            .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
            .map(|d| d.as_millis() as f64)
            .unwrap_or(0.0);
        if modified_ms < cutoff_ms {
            continue;
        }
        let session_id = fname.trim_end_matches(".jsonl").to_string();
        scanned_sessions += 1;
        let _ = scan_session(&entry.path(), &session_id, &mut accumulator);
    }

    let mut hotspots: Vec<RepeatedToolError> = accumulator
        .into_iter()
        .filter_map(|((tool_name, error_prefix), acc)| {
            if acc.occurrences < min_occurrences {
                return None;
            }
            let mut session_ids: Vec<String> = acc.sessions.into_iter().collect();
            session_ids.sort();
            Some(RepeatedToolError {
                tool_name,
                error_prefix,
                occurrences: acc.occurrences,
                session_count: session_ids.len() as u32,
                session_ids,
                last_seen_ms: acc.last_seen_ms,
            })
        })
        .collect();

    hotspots.sort_by(|a, b| b.occurrences.cmp(&a.occurrences));

    Ok(ErrorHotspotsResponse {
        repeated_errors: hotspots,
        scanned_sessions,
    })
}
